src/test/java/apoc/nodes/NodesTest.java

```diff
@@ -277,4 +281,233 @@ public void testRelType() {
+                            Util.map("name", "Neo"), asSet(label("Company"))); //TO
+                    assertFalse(r.hasNext());
+                });
+    }
+
+    @Test
+    public void testMergeVirtualNode() {
+        db.execute("CREATE \n" +
+                "(p:Person {name: 'John'})-[:LIVES_IN]->(c:City{name:'London'}),\n" +
+                "(p1:Person {name: 'Mike'})-[:LIVES_IN]->(c),\n" +
+                "(p2:Person {name: 'Kate'})-[:LIVES_IN]->(c),\n" +
+                "(p3:Person {name: 'Budd'})-[:LIVES_IN]->(c),\n" +
+                "(p4:Person {name: 'Alex'})-[:LIVES_IN]->(c),\n" +
+                "(p1)-[:KNOWS]->(p),\n" +
+                "(p2)-[:KNOWS]->(p1),\n" +
+                "(p2)-[:KNOWS]->(p3),\n" +
+                "(p4)-[:KNOWS]->(p3)\n").close();
+
+        Set<Label> label = asSet(label("City"), label("Person"));
+
+        TestUtil.testResult(db, "MATCH (p:Person)-[:LIVES_IN]->(c:City)\n" +
+                "WITH c, c + collect(p) as subgraph\n" +
+                "CALL apoc.nodes.collapse(subgraph,{mergeVirtualRels:true, countMerge: true}) yield from, rel, to return from, rel, to", null, result -> {
+            Map<String, Object> map = result.next();
+
+            assertEquals(Util.map("name","John", "count", 6), ((VirtualNode)map.get("from")).getAllProperties());
+            assertEquals(label, asSet(((VirtualNode) map.get("from")).getLabels().iterator()));
+            assertNull(((VirtualRelationship) map.get("rel")));
+            assertNull(((VirtualNode) map.get("to")));
+            assertFalse(result.hasNext());
+        });
+    }
+
+    @Test
+    public void testMergeVirtualNodeBOTH() {
+        db.execute("CREATE \n" +
+                "(p:Person {name: 'John'})-[:LIVES_IN]->(c:City{name:'London'})," +
+                "(c)-[:LIVES_IN]->(p)").close();
+
+        Set<Label> label = asSet(label("City"), label("Person"));
+
+        TestUtil.testResult(db, "MATCH (p:Person)-[:LIVES_IN]->(c:City)-[:LIVES_IN]->(b:Person)\n" +
+                "CALL apoc.nodes.collapse([c,p,b],{mergeVirtualRels:true, countMerge: true, selfRel: true}) yield from, rel, to return from, rel, to", null, result -> {
+            Map<String, Object> map = result.next();
+
+            assertMerge(map,
+                    Util.map("name", "John", "count", 2), label, //FROM
+                    Util.map("count", 3), "LIVES_IN", //REL
+                    Util.map("name", "John", "count", 2), label); //TO
+            assertFalse(result.hasNext());
+        });
+    }
+
+    private static void assertMerge(Map<String, Object> map,
+                                    Map<String, Object> fromProperties, Set<Label> fromLabel,
+                                    Map<String, Object> relProperties, String relType,
+                                    Map<String, Object> toProperties, Set<Label> toLabel
+    ) {
+        assertEquals(fromProperties, ((VirtualNode)map.get("from")).getAllProperties());
```

Comment:
Estrai `(VirtualNode)map.get("from")` e `(VirtualRelationship)map.get("rel")` in delle variabili